### ISSUE TITLE
BUG: read_csv returns inconsistent or misleading values with dtype=float

### ISSUE THREAD

**Issue by c06n** [pandas-dev/pandas#34120]
The issue occurs for `read_csv()` when reading a mixture of empty or NaN and potential Boolean values when choosing `float` as dtype. NaN is either read as 1.0, or in the case of an empirical csv-file even inconsistently either as 1.0 or NaN. I could not replicate the inconsistency with synthetic-data, therefore the csv-file in question is provided here: https://github.com/c06n/Pandas_readcsv_issue/blob/master/empirical_data.csv 

Neither issue occurs if `engine='python'`, therefore I believe it is a bug.

#### example with synthetic data:

```python
import pandas as pd
import numpy as np

df = pd.DataFrame({'a': np.r_[[np.nan]*2000,
                              np.repeat('True', 100),
                              [np.nan]*2000,
                              np.repeat('True', 100),
                              [np.nan]*2000]})
df.to_csv('synthetic_data.csv', sep=';', decimal=',', index=False)
df = pd.read_csv('synthetic_data.csv', 
                 na_values='',
                 sep=';',
                 decimal=',',
                 dtype={'a': 'float'})
df.a
```

#### example with empirical data:

Get the data from https://github.com/c06n/Pandas_readcsv_issue/blob/master/empirical_data.csv and then use the code below. The interesting columns are the right-most ones, e.g. the column `approach`. 

```
df = pd.read_csv('empirical_data.csv', 
                 na_values='',
                 sep=';',
                 decimal=',',
                 dtype={'approach': 'float'})
df.approach
```

#### Problem description

In the synthetic data, all values are converted to 1.0 (both NaN-values and the `True`-string object). In the empirical data, the first NaN-values are converted to 1.0, but at the end there are values that remain NaN.

* The problematic nature of the inconsistent treatment of NaN in case of the empirical data file should be self explanatory. 
* Returning 1.0 for NaN-values seems inappropriate as well. Because the Python-engine returns NaN, this behavior does not appear to be intended.

Personally, I find the behavior in either case hugely problematic, because converting NaN to 1.0 is very non-intuitive in either case. Throwing an error, if possible, would be my preference.

#### Expected Output

Empty cells and values that are explicitely defined as NaN (e.g. "nan") should remain NaN.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.7.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 142 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.None

pandas           : 1.0.3
numpy            : 1.18.1
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 46.1.3.post20200330
Cython           : 0.29.17
pytest           : None
hypothesis       : 5.11.0
sphinx           : 3.0.3
blosc            : None
feather          : None
xlsxwriter       : 1.2.8
lxml.etree       : 4.5.0
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.13.0
pandas_datareader: None
bs4              : 4.9.0
bottleneck       : 1.3.2
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.5.0
matplotlib       : 3.1.3
numexpr          : 2.7.1
odfpy            : None
openpyxl         : 3.0.3
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : None
pyxlsb           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : 1.3.16
tables           : 3.6.1
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : 1.3.0
xlsxwriter       : 1.2.8
numba            : 0.49.0

</details>

**Comment by hasB4K** [comment]
Hello,

I have reproduced the error with the following snippet:

```python
import pandas as pd
import os

def generate_buggy_file(n_cols, **extend_read_csv):
    with open("tmp.csv", "wt") as tmp_csv:
        tmp_csv.write("test;\n")
        tmp_csv.write("True;\n")
        for i in range(n_cols):
            tmp_csv.write(";\n")
            
    print("tmp.csv - size (KBi):", os.path.getsize("tmp.csv") / 1024)
    
    return pd.read_csv('tmp.csv', sep=";", **extend_read_csv)

print("==> Small buggy file (only 1.0)")
print(generate_buggy_file(4, dtype={"test": "float"}).test)

print("==> Small buggy file (no dtype)")
print(generate_buggy_file(4).test)

print("\n==> Small buggy file (correct: engine=python)")
print(generate_buggy_file(4, dtype={"test": "float"}, engine="python").test)

print("\n==> Big buggy file (issue just after 512KBi)")
print(generate_buggy_file(2**18, dtype={"test": "float"}).test)
```

Outputs:
```
==> Small buggy file (only 1.0)
tmp.csv - size (KBi): 0.01953125
0    1.0
1    1.0
2    1.0
3    1.0
4    1.0
Name: test, dtype: float64
==> Small buggy file (no dtype)
tmp.csv - size (KBi): 0.01953125
0    True
1     NaN
2     NaN
3     NaN
4     NaN
Name: test, dtype: object

==> Small buggy file (correct: engine=python)
tmp.csv - size (KBi): 0.01953125
0    1.0
1    NaN
2    NaN
3    NaN
4    NaN
Name: test, dtype: float64

==> Big buggy file (issue just after 512KBi)
tmp.csv - size (KBi): 512.01171875
0         1.0
1         1.0
2         1.0
3         1.0
4         1.0
         ... 
262140    1.0
262141    1.0
262142    1.0
262143    1.0
262144    NaN
Name: test, Length: 262145, dtype: float64
```

TL;DR:
* The column `approach` in the `empirical_data.csv` is either `'True'` or empty, I simulate this with `generate_buggy_file`
* It only happens if we specify the dtype
* It seems that the bug is only happening with CSV larger than 512KBi which is equal to the `DEFAULT_CHUNKSIZE` (2**18) here: https://github.com/pandas-dev/pandas/blob/9929fca82f305cd5388f235b4abd402b0f2e3653/pandas/_libs/parsers.pyx#L257
